README.md
[![Build Status](https://secure.travis-ci.org/ask11/backbone-id.png?branch=master)](https://travis-ci.org/ask11/backbone-id)

# Backbone.Id

  Generates Id for new instance of Backbone.Model on the client.
  Sometimes it useful to generate model's id imediately, in order to create relationships with other models and generate views. Supports [UUID](http://en.wikipedia.org/wiki/Universally_unique_identifier), [MongoDB ObjectId](http://docs.mongodb.org/manual/core/object-id/) and custom generators for your taste and needs.
  If you want to generate timestamps fields (createdAt, updatedAt) on the client, use [backbone-timestamp](https://github.com/ask11/backbone-timestamp).

## Example

```js
// we have model
var Book = Backbone.Model.extend({
  idAttribute: '_id',
  defaults: {
    title: 'unknown book',
    author: 'unknown'
  }
});

// apply Backbone.Id to Book model
Backbone.Id(Book);

var book = new Book({ title: 'The Lord of the Rings' });
book.toJSON();
// => {
//  title: "The Lord of the Rings",
//  author: "unknown",
//  _id: "9f08362e-dbbd-4675-9ff7-f96c9726f29f"
// }

book.isNew(); // => true
```

## API - Backbone.id(Model, method)

  Plugin adds one `Backbone.Id` function to patch existing Backbone.Model with id-generator functionality.
  By default it uses UUID, and shiped with MongoDB's ObjectId generator

```js
Backbone.Id(Book, 'mongo');
var book = new Book({ title: 'Red' });
book.toJSON(); // => { title: "Red", author: "unknown", _id: "305fd868a7627428df000002" }
```

  Also it supports custom function as a generator.

```js
var iterator = 101;
Backbone.Id(Book, function() {
  return iterator++;
});

model = new Book({ title: 'Effective Javascript' });
model.id; // => 101

model = new Book({ title: 'Node.js in action' });
model.id; // => 102
```

## Development

  - `npm install` to install development dependencies: bower & testem.
  - `npm test` to ensure that all tests pass
  - `npm start` to run test watcher on [localhost:7357](http://localhost:7357/)

## Licence

  (c) 2013, Aleksey Kulikov
  May be freely distributed according to MIT license.


History.md
## 0.3.0 / 2013-07-..

  * update Reamde with isNew explanation
  * update bower to 0.10.0
  * update underscore to 1.5.0

### 0.2.3 / 2013-06-20

  * add `Backbone.Id.guid` and `Backbone.Id.mongo` as a standalone id generators
  * updated bower and testem configs

### 0.2.2 / 2013-06-10

  * use index.js as a main file
  * notifications for .travis.yml
  * use mit-license.org as LICENCE source

### 0.2.1 / 2013-06-04

  * migration to bower as a package manager

## 0.2.0 / 2013-05-27

  - fixed bugs with `isNew`
  - fixed bug with `initialize` method
  - added Readme
  - use travis-ci
  - improved dev requirements

## 0.1.0 / 2013-03-11

  - initial release


test/index-test.js
describe('Backbone.Id', function() {
  var expect = chai.expect;
  var Book = Backbone.Model.extend({
    idAttribute: '_id',
    defaults: {
      title: 'unknown book',
      author: 'unknown'
    }
  });

  Backbone.Id(Book);
  Backbone.sync = function() {};

  it('has #mongo and #guid generators', function() {
    expect(Backbone.Id.mongo()).length(24);
    expect(Backbone.Id.guid()).length(36);
  });

  it('generates uuid-like id automatically', function() {
     var book = new Book({ title: 'The Lord of the Rings' });

     expect(book.get('title')).equal('The Lord of the Rings');
     expect(book.get('author')).equal('unknown');

     expect(book.get('_id')).length(36);
     expect(book.id).match(/^[a-f,0-9]{8}\-[a-f,0-9]{4}\-4[a-f,0-9]{3}\-[a-b,8-9][a-f,0-9]{3}\-[a-f,0-9]{12}$/);
  });

  it('generates mongo-like id', function() {
    var Model = Backbone.Model.extend({});
    Backbone.Id(Model, 'mongo');

    var model = new Model({ name: 'test mongo id generator' });
    expect(model.id).length(24);
    expect(model.id).match(/[a-f,0-9]{24}/);
  });

  it('allows to use custom id function', function() {
    var iterator = 1, model;
    var Model = Backbone.Model.extend({});
    Backbone.Id(Model, function() {
      return iterator++;
    });

    _.times(100, function() {
      model = new Model();
    });

    expect(model.id).equal(100);
  });

  it('isNew works as expect', function() {
    var book = new Book({ title: 'The Hobbit' });
    expect(book.isNew()).equal(true);
  });

  it('removes _new pseudo attribute on first sync', function() {
    var Model = Backbone.Model.extend({
      sync: function(method, model, options) {
        return expect(method).equal('create');
      }
    });

    Backbone.Id(Model);
    var model = new Model({ name: 'test' });
    model.save();

    expect(_.values(model.attributes)).length(2);
    expect(model.isNew()).equal(false);
  });

  it('does not change existing id', function() {
    var book = new Book({ _id: 1, title: 'A Tale of Two Cities' });
    expect(book.id).equal(1);
  });

  it('does not add _new when id exists', function() {
    var book = new Book({ _id: 'asd', title: 'Romeo & Juliet' });
    expect(book.has('_new')).equal(false);
    expect(book.isNew()).equal(false);
  });

  it('generates id on collection.create', function () {
    var Library = Backbone.Collection.extend({
      model: Book
    });

    var library = new Library();
    library.create({ title: 'And Then There Were None' });
    expect(library.at(0).id).length(36);
  });
});


test/backbone-id.test.js
describe('Backbone.Id', function() {
  var expect = chai.expect;
  var Book = Backbone.Model.extend({
    idAttribute: '_id',
    defaults: {
      title: 'unknown book',
      author: 'unknown'
    }
  });

  before(function() {
    Backbone.Id(Book);
    Backbone.sync = function() {};
  });

  it('generates uuid-like id automatically', function() {
     var book = new Book({ title: 'The Lord of the Rings' });

     expect(book.get('title')).equal('The Lord of the Rings');
     expect(book.get('author')).equal('unknown');

     expect(book.get('_id')).length(36);
     expect(book.id).match(/^[a-f,0-9]{8}\-[a-f,0-9]{4}\-4[a-f,0-9]{3}\-[a-b,8-9][a-f,0-9]{3}\-[a-f,0-9]{12}$/);
  });

  it('generates mongo-like id', function() {
    var Model = Backbone.Model.extend({});
    Backbone.Id(Model, 'mongo');

    var model = new Model({ name: 'test mongo id generator' });
    expect(model.id).length(24);
    expect(model.id).match(/[a-f,0-9]{24}/);
  });

  it('allows to use custom id function', function() {
    var iterator = 1, model;
    var Model = Backbone.Model.extend({});
    Backbone.Id(Model, function() {
      return iterator++;
    });

    _.times(100, function() {
      model = new Model();
    });

    expect(model.id).equal(100);
  });

  it('isNew works as expect', function() {
    var book = new Book({ title: 'The Hobbit' });
    expect(book.isNew()).equal(true);
  });

  it('removes _new pseudo attribute on first sync', function() {
    var Model = Backbone.Model.extend({
      sync: function(method, model, options) {
        return expect(method).equal('create');
      }
    });

    Backbone.Id(Model);
    var model = new Model({ name: 'test' });
    model.save();

    expect(_.values(model.attributes)).length(2);
    expect(model.isNew()).equal(false);
  });

  it('does not change existing id', function() {
    var book = new Book({ _id: 1, title: 'A Tale of Two Cities' });
    expect(book.id).equal(1);
  });

  it('does not add _new when id exists', function() {
    var book = new Book({ _id: 'asd', title: 'Romeo & Juliet' });
    expect(book.has('_new')).equal(false);
    expect(book.isNew()).equal(false);
  });

  it('generates id on collection.create', function () {
    var Library = Backbone.Collection.extend({
      model: Book
    });

    var library = new Library();
    library.create({ title: 'And Then There Were None' });
    expect(library.at(0).id).length(36);
  });
});


test/backbone_id_test.js
describe('Backbone.Id', function() {
  var expect = chai.expect;
  var Book = Backbone.Model.extend({
    idAttribute: '_id',
    defaults: {
      title: 'unknown book',
      author: 'unknown'
    }
  });
  Backbone.Id(Book);
  Backbone.sync = function() {};

  it('generates uuid-like id automatically', function() {
     var book = new Book({ title: 'The Lord of the Rings' });

     expect(book.get('title')).equal('The Lord of the Rings');
     expect(book.get('author')).equal('unknown');

     expect(book.id).exist;
     expect(book.get('_id')).length(36);
     expect(book.id).match(/^[a-f,0-9]{8}\-[a-f,0-9]{4}\-4[a-f,0-9]{3}\-[a-b,8-9][a-f,0-9]{3}\-[a-f,0-9]{12}$/);
  });

  it('generates mongo-like id', function() {
    var Model = Backbone.Model.extend({});
    Backbone.Id(Model, 'mongo');

    var model = new Model({ name: 'test mongo id generator' });
    expect(model.id).length(24);
    expect(model.id).match(/[a-f,0-9]{24}/);
  });

  it('allows to use custom id function', function() {
    var iterator = 1, model;
    var Model = Backbone.Model.extend({});
    Backbone.Id(Model, function() {
      return iterator++;
    });

    _.times(100, function() {
      model = new Model();
    });

    expect(model.id).equal(100);
  });

  it('isNew works as expect', function() {
    var book = new Book({ title: 'The Hobbit' });
    expect(book.isNew()).equal(true);
  });

  it('removes _new pseudo attribute on first sync', function() {
    var Model = Backbone.Model.extend({
      sync: function(method, model, options) {
        return expect(method).equal('create');
      }
    });

    Backbone.Id(Model);
    var model = new Model({ name: 'test' });
    model.save();

    expect(_.values(model.attributes)).length(2);
    expect(model.isNew()).equal(false);
  });

  it('does not change existing id', function() {
    var book = new Book({ _id: 1, title: 'A Tale of Two Cities' });
    expect(book.id).equal(1);
  });

  it('generates id on collection.create', function () {
    var Library = Backbone.Collection.extend({
      model: Book
    });

    var library = new Library();
    library.create({ title: 'And Then There Were None' });
    expect(library.at(0).id).length(36);
  });
});


index.js
// Id generator for Backbone.Model
// https://github.com/ask11/backbone-id

;(function(Backbone, _) {
  'use strict';

  // Implementation of ObjectId generator
  // http://docs.mongodb.org/manual/core/object-id/
  var increment = 0x1000000;
  var localId1  = ((1+Math.random())*0x100000 | 0).toString(16).substring(1);
  var localId2  = ((1+Math.random())*0x100000 | 0).toString(16).substring(1);

  function mongo() {
    var dateNow = ((new Date()).getTime()/100 | 0).toString(16);
    return dateNow + localId1 + localId2 + (++increment).toString(16).substring(1);
  }

  // http://stackoverflow.com/a/2117523/893744
  // About UUID format https://gist.github.com/jed/982883#file-readme-md
  function guid() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
      var r = Math.random()*16|0, v = c == 'x' ? r : (r&0x3|0x8);
      return v.toString(16);
    });
  }

  Backbone.Id = function(Model, method) {
    var idAttribute = Model.prototype.idAttribute;
    var defaults    = Model.prototype.defaults;
    var sync        = Model.prototype.sync;
    var initialize  = Model.prototype.initialize;

    if (method === 'mongo') method = mongo;
    if (!method || method === 'guid') method = guid;

    Model.prototype.defaults = function() {
      var defaultAttrs = _.isFunction(defaults) ? defaults() : defaults || {};
      var id           = method();
      var attrs        = {};

      this._new = id;
      attrs[idAttribute] = id;
      return _.defaults({}, defaultAttrs, attrs);
    };

    Model.prototype.initialize = function() {
      if (this.get(idAttribute) !== this._new) delete this._new;
      return initialize.apply(this, arguments);
    };

    Model.prototype.isNew = function() {
      return !! this._new;
    };

    Model.prototype.sync = function() {
      var defer = sync.apply(this, arguments);
      if (this.isNew()) delete this._new;
      return defer;
    };
  };

  /// expose id generators
  Backbone.Id.mongo = mongo;
  Backbone.Id.guid  = guid;
}).call(this, Backbone, _);


backbone-id.js
// Id generator for Backbone.Model
// https://github.com/ask11/backbone-id
//
// (c) 2013, Aleksey Kulikov
// May be freely distributed according to MIT license.

;(function(Backbone, _) {
  'use strict';

  // Implementation of ObjectId generator
  // http://docs.mongodb.org/manual/core/object-id/
  var increment = 0x1000000;
  var localId1  = ((1+Math.random())*0x100000 | 0).toString(16).substring(1);
  var localId2  = ((1+Math.random())*0x100000 | 0).toString(16).substring(1);

  function mongo() {
    var dateNow = ((new Date()).getTime()/100 | 0).toString(16);
    return dateNow + localId1 + localId2 + (++increment).toString(16).substring(1);
  }

  // http://stackoverflow.com/a/2117523/893744
  // About UUID format https://gist.github.com/jed/982883#file-readme-md
  function guid() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
      var r = Math.random()*16|0, v = c == 'x' ? r : (r&0x3|0x8);
      return v.toString(16);
    });
  }

  Backbone.Id = function(Model, method) {
    var idAttribute = Model.prototype.idAttribute;
    var defaults    = Model.prototype.defaults;
    var sync        = Model.prototype.sync;
    var initialize  = Model.prototype.initialize;

    if (method === 'mongo') method = mongo;
    if (!method || method === 'guid') method = guid;

    Model.prototype.defaults = function() {
      var defaultAttrs = _.isFunction(defaults) ? defaults() : defaults || {};
      var id           = method();
      var attrs        = {};

      this._new = id;
      attrs[idAttribute] = id;
      return _.defaults({}, defaultAttrs, attrs);
    };

    Model.prototype.initialize = function() {
      if (this.get(idAttribute) !== this._new) delete this._new;
      return initialize.apply(this, arguments);
    };

    Model.prototype.isNew = function() {
      return !! this._new;
    };

    Model.prototype.sync = function() {
      var defer = sync.apply(this, arguments);
      if (this.isNew()) delete this._new;
      return defer;
    };
  };
}).call(this, Backbone, _);


backbone.id.js
// ...
// https://github.com/Ask11/backbone.id
//
// (c) 2013, Aleksey Kulikov
// May be freely distributed according to MIT license.

;(function(Backbone, _) {
  'use strict';

  // Basic implementation of ObjectId generator
  // http://docs.mongodb.org/manual/core/object-id/
  var increment = 0x1000000
    , localId1  = ((1+Math.random())*0x100000 | 0).toString(16).substring(1)
    , localId2  = ((1+Math.random())*0x100000 | 0).toString(16).substring(1);

  function mongo() {
    var dateNow = ((new Date()).getTime()/100 | 0).toString(16);
    return dateNow + localId1 + localId2 + (++increment).toString(16).substring(1);
  }

  // http://stackoverflow.com/a/2117523/893744
  // About UUID format https://gist.github.com/jed/982883#file-readme-md
  function guid() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
      var r = Math.random()*16|0, v = c == 'x' ? r : (r&0x3|0x8);
      return v.toString(16);
    });
  }

  Backbone.Id = function(Model, method) {
    var idAttribute = Model.prototype.idAttribute
      , defaults    = Model.prototype.defaults
      , sync        = Model.prototype.sync;

    if (method == 'mongo') method = mongo;
    if (!method || method == 'guid') method = guid;

    Model.prototype.defaults = function() {
      var defaultValues = _.isFunction(defaults) ? defaults() : defaults || {}
        , backboneId    = { _new: true };

      backboneId[idAttribute] = method();
      return _.defaults(defaultValues, backboneId);
    };

    Model.prototype.isNew = function() {
      return !!this.get('_new');
    };

    Model.prototype.sync = function() {
      var defer = sync.apply(this, arguments);
      if (this.isNew()) this.unset('_new');
      return defer;
    };
  };
}).call(this, Backbone, _);
